src/Microsoft.TemplateEngine.Edge/Template/ParameterSetBuilder.cs

```diff
@@ -0,0 +1,361 @@
+            }
+
+            paramsWithInvalidValues = paramsWithInvalidValuesList;
+            return templateParams;
+        }
+
+        public bool TryGetParameterDefinition(string name, out ITemplateParameter parameter) => TryGetValue(name, out parameter);
+
+        public void SetParameterValue(ITemplateParameter parameter, object value, DataSource dataSource)
+        {
+            _resolvedValues[parameter].SetValue(value, dataSource);
+            _result = null;
+        }
+
+        public void SetParameterEvaluation(ITemplateParameter parameter, EvaluatedInputParameterData evaluatedParameterData)
+        {
+            var old = _resolvedValues[parameter];
+            _resolvedValues[parameter] = new EvalData(evaluatedParameterData);
+            if (old.InputDataState != InputDataState.Unset)
+            {
+                _resolvedValues[parameter].SetValue(old.Value, old.DataSource);
+            }
+
+            _result = null;
+        }
+
+        public bool HasParameterValue(ITemplateParameter parameter) => _resolvedValues[parameter].InputDataState != InputDataState.Unset;
+
+        public bool CheckIsParametersEvaluationCorrect(IGenerator generator, ILogger logger, out IReadOnlyList<string> paramsWithInvalidEvaluations)
+        {
+            List<EvalData> evaluatedParameters = _resolvedValues.Values.ToList();
+            List<EvalData> clonedParameters = evaluatedParameters.Select(v => v.Clone()).ToList();
+            RunDatasetEvaluation(clonedParameters, generator, logger);
+            List<string> invalidParams = new List<string>();
+
+            foreach (var pair in evaluatedParameters.Zip(clonedParameters, (a, b) => (a, b)))
+            {
+                if (pair.a.IsEnabledConditionResult != pair.b.IsEnabledConditionResult ||
+                    pair.a.IsRequiredConditionResult != pair.b.IsRequiredConditionResult)
+                {
+                    invalidParams.Add(pair.a.ParameterDefinition.Name);
+                }
+            }
+
+            paramsWithInvalidEvaluations = invalidParams;
+            return invalidParams.Count == 0;
+        }
+
+        public void EvaluateConditionalParameters(IGenerator generator, ILogger logger)
+        {
+            List<EvalData> evaluatedParameters = _resolvedValues.Values.ToList();
+            RunDatasetEvaluation(evaluatedParameters, generator, logger);
+        }
+
+        public IEvaluatedInputDataSet Build()
+        {
+            if (_result == null)
+            {
+                _result = new EvaluatedInputDataSet(
+                    this,
+                    _resolvedValues.Select(p => p.Value.ToParameterData()).ToList());
+            }
+
+            return _result!;
+        }
+
+        internal static void SetParameterDefault(IGenerator generator, IParameterSetBuilder templateParams, ITemplateParameter parameter, IEngineEnvironmentSettings environment, bool useHostDefaults, bool isRequired, List<string> paramsWithInvalidValues)
```

Comment:
```suggestion
        internal static void SetParameterDefault(IGenerator generator, IParameterSetBuilder templateParams, ITemplateParameter parameter, IEngineEnvironmentSettings environment, bool useHostDefaults, bool isRequired, out List<string> paramsWithInvalidValues)
```

Reply:
Would it be better to have this method as non-static? Can we avoid using `isRequired` if the method is non-static? it looks like `ParameterSetBuilder` should already know if the parameter is required or not.